sdk/identity/azure-identity/src/main/java/com/azure/identity/ClientCertificateCredential.java

```diff
@@ -40,22 +38,19 @@
     ClientCertificateCredential(String tenantId, String clientId, String certificatePath, String certificatePassword,
                                 IdentityClientOptions identityClientOptions) {
         Objects.requireNonNull(certificatePath, "'certificatePath' cannot be null.");
-        this.clientCertificate = certificatePath;
-        this.clientCertificatePassword = certificatePassword;
-        identityClient =
-            new IdentityClientBuilder()
-                .tenantId(tenantId)
-                .clientId(clientId)
-                .identityClientOptions(identityClientOptions)
-                .build();
+        identityClient = new IdentityClientBuilder()
+            .tenantId(tenantId)
+            .clientId(clientId)
+            .certificatePath(certificatePath)
+            .certificatePassword(certificatePassword)
+            .identityClientOptions(identityClientOptions)
+            .build();
     }
 
     @Override
     public Mono<AccessToken> getToken(TokenRequestContext request) {
-        if (clientCertificatePassword != null) {
-            return identityClient.authenticateWithPfxCertificate(clientCertificate, clientCertificatePassword, request);
-        } else {
-            return identityClient.authenticateWithPemCertificate(clientCertificate, request);
-        }
+        return identityClient.authenticateWithConfidentialClientCache(request)
+            .onErrorResume(t -> Mono.empty())
+            .switchIfEmpty(Mono.defer(() -> identityClient.authenticateWithConfidentialClient(request)));
```

Comment:
I don't know if we need the `Mono.defer` here, I believe `switchIfEmpty` only processes if the upstream returns empty otherwise it is never ran.

Reply:
switchIfEmpty only processes the Mono if the upstream returns empty, but the method call `identityClient.authenticateWithConfidentialClient()` would be executed before the stream starts.